### PR TITLE
[BROWSEUI] Implement CLSID_ACListISF

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose

Implement enumeration of `IShellFolder` items of auto-completion.

JIRA issue: [CORE-9281](https://jira.reactos.org/browse/CORE-9281)

BEFORE:
before
There were many failures.

AFTER:
after
Successful.

### CODE REVIEW

File: dll/win32/browseui/aclistisf.cpp [hunk 1]
```diff
@@ -20,83 +21,411 @@
...
+            return S_FALSE;
+    }
+    return S_OK;
 }
 
-HRESULT STDMETHODCALLTYPE CACListISF::Skip(ULONG celt)
+STDMETHODIMP CACListISF::Skip(ULONG celt)
 {
-    TRACE("(%p, %d)\n", this, celt);
+    TRACE("Skip(%p, %d)\n", this, celt);
```
Comment (learn-more) [inline dll/win32/browseui/aclistisf.cpp:325]: If you set the environment variable `DEBUGFORMAT=wine` the function name will be included.

File: dll/win32/browseui/aclistisf.cpp [hunk 2]
```diff
@@ -20,83 +21,411 @@
...
+                }
+            }
+            // FALL THROUGH
+        case LT_DESKTOP:
+            ++m_iNextLocation;
+            if (m_dwOptions & ACLO_DESKTOP)
+            {
+                CComHeapPtr<ITEMIDLIST> pidl;
+                hr = SHGetSpecialFolderLocation(NULL, CSIDL_DESKTOP, &pidl);
+                if (FAILED(hr))
```
Comment (learn-more) [inline dll/win32/browseui/aclistisf.cpp:59]: FAILED_UNEXPECTEDLY?

File: dll/win32/browseui/aclistisf.cpp [hunk 3]
```diff
@@ -20,83 +21,411 @@
...
+                }
+            }
+            // FALL THROUGH
+        case LT_MYCOMPUTER:
+            ++m_iNextLocation;
+            if (m_dwOptions & ACLO_MYCOMPUTER)
+            {
+                CComHeapPtr<ITEMIDLIST> pidl;
+                hr = SHGetSpecialFolderLocation(NULL, CSIDL_DRIVES, &pidl);
+                if (FAILED(hr))
```
Comment (learn-more) [inline dll/win32/browseui/aclistisf.cpp:75]: FAILED_UNEXPECTEDLY?

File: dll/win32/browseui/aclistisf.h [hunk 4]
```diff
@@ -25,40 +26,68 @@ class CACListISF :
...
+    {
+        LT_DIRECTORY,
+        LT_DESKTOP,
+        LT_MYCOMPUTER,
+        LT_FAVORITES,
+        LT_MAX
+    };
+
     DWORD m_dwOptions;
+    INT m_iNextLocation; /* LOCATION_TYPE */
```
Comment (learn-more) [inline dll/win32/browseui/aclistisf.h:44]: Why INT here, and not directly LOCATION_TYPE?

File: dll/win32/browseui/aclistisf.cpp [hunk 5]
```diff
@@ -20,83 +21,411 @@
...
+    TRACE("SetLocation(%p, %p)\n", this, pidl);
+
+    m_pEnumIDList.Release();
+    m_pShellFolder.Release();
+    m_pidlLocation.Free();
+
+    if (!pidl)
+        return E_FAIL;
+
+    LPITEMIDLIST pidlClone = ILClone(pidl);
```
Comment (learn-more) [inline dll/win32/browseui/aclistisf.cpp:119]: Can use the m_pidlLocation directly here (unless there is a good reason not to?)

File: dll/win32/browseui/aclistisf.cpp [hunk 6]
```diff
@@ -20,83 +21,411 @@
...
+        m_pShellFolder.Attach(pFolder.Detach());
+
+    SHCONTF Flags = SHCONTF_FOLDERS | SHCONTF_INIT_ON_FIRST_NEXT;
+    if (m_fShowHidden)
+        Flags |= SHCONTF_INCLUDEHIDDEN;
+    if (!(m_dwOptions & ACLO_FILESYSDIRS))
+        Flags |= SHCONTF_NONFOLDERS;
+
+    hr = m_pShellFolder->EnumObjects(NULL, Flags, &m_pEnumIDList);
+    if (hr != S_OK)
```
Comment (learn-more) [inline dll/win32/browseui/aclistisf.cpp:147]: can this be S_FALSE?
Do you need to report that as E_FAIL?

File: dll/win32/browseui/aclistisf.cpp [hunk 7]
```diff
@@ -20,83 +21,411 @@
...
+
+HRESULT CACListISF::GetDisplayName(LPCITEMIDLIST pidlChild, CComHeapPtr<WCHAR>& pszChild)
+{
+    TRACE("GetDisplayName(%p, %p)\n", this, pidlChild);
+    pszChild.Free();
+
+    STRRET StrRet;
+    DWORD dwFlags = SHGDN_INFOLDER | SHGDN_FORPARSING | SHGDN_FORADDRESSBAR;
+    HRESULT hr = m_pShellFolder->GetDisplayNameOf(pidlChild, dwFlags, &StrRet);
+    if (FAILED(hr))
```
Comment (learn-more) [inline dll/win32/browseui/aclistisf.cpp:164]: Does this happen, is this expected?

File: dll/win32/browseui/aclistisf.cpp [hunk 8]
```diff
@@ -20,83 +21,411 @@
...
+    {
+        dwFlags = SHGDN_INFOLDER | SHGDN_FORPARSING;
+        hr = m_pShellFolder->GetDisplayNameOf(pidlChild, dwFlags, &StrRet);
+        if (FAILED(hr))
```
Comment (learn-more) [inline dll/win32/browseui/aclistisf.cpp:168]: FAILED_UNEXPECTEDLY?
Same for all cases that are not expected to fail.

**Review by learn-more — COMMENTED** [review]
Some minor comments, mainly related to easier debugging corner cases / unexpected errors in other components.

File: dll/win32/browseui/aclistisf.cpp (hunk 1 above)
Comment (katahiromz) [reply dll/win32/browseui/aclistisf.cpp]: The function names in TRACE are removed. OK.

File: dll/win32/browseui/aclistisf.cpp (hunk 2 above)
Comment (katahiromz) [reply dll/win32/browseui/aclistisf.cpp]: OK.

File: dll/win32/browseui/aclistisf.cpp (hunk 3 above)
Comment (katahiromz) [reply dll/win32/browseui/aclistisf.cpp]: OK.

File: dll/win32/browseui/aclistisf.h (hunk 4 above)
Comment (katahiromz) [reply dll/win32/browseui/aclistisf.h]: OK.

File: dll/win32/browseui/aclistisf.cpp (hunk 8 above)
Comment (katahiromz) [reply dll/win32/browseui/aclistisf.cpp]: OK.

File: dll/win32/browseui/aclistisf.cpp (hunk 6 above)
Comment (katahiromz) [reply dll/win32/browseui/aclistisf.cpp:144]: https://docs.microsoft.com/en-us/windows/win32/api/shobjidl_core/nf-shobjidl_core-ishellfolder-enumobjects

> Some implementations may also return S_FALSE, indicating that there are no children matching the grfFlags that were passed in. If S_FALSE is returned, ppenumIDList is set to NULL.

File: dll/win32/browseui/aclistisf.cpp (hunk 7 above)
Comment (katahiromz) [reply dll/win32/browseui/aclistisf.cpp:161]: `SHGDN_FORADDRESSBAR` might fail in some situation.

File: dll/win32/browseui/aclistisf.cpp (hunk 5 above)
Comment (katahiromz) [reply dll/win32/browseui/aclistisf.cpp]: OK.

File: dll/win32/browseui/aclistisf.cpp (hunk 6 above)
Comment (katahiromz) [reply dll/win32/browseui/aclistisf.cpp:145]: `SetLocation` is used in `Reset`, `Expand`, and `NextLocation`. They expect `S_OK` or failure of `SetLocation`. `S_FALSE` of `EnumObjects` should treat as a failure.

**Pull request merged**